### ISSUE TITLE
False positive on multi-arity Rum stateful components

### ISSUE THREAD

**Issue by bfontaine** [borkdude/clj-kondo#987]
**version**

v2020.07.29

**macro usage**

**platform**

native

**problem**

The [example code](https://github.com/borkdude/clj-kondo/blob/master/examples/rum/.clj-kondo/hooks/rum.clj) to deal with Rum’s `defcs` can’t parse multi-arity components: it raises false-positive arity issues.

**repro**

```
├── .clj-kondo
│   ├── config.edn
│   └── hooks
│       └── rum.clj
└── repro.clj
```

```clojure
;; repro.clj
(ns repro
  (:require [rum.core :as rum]))

(rum/defcs compo < rum/static
  ([state a]
   [:p (pr-str state a)])
  ([state a b]
   [:p (pr-str state a b)]))

(compo 42)
```

```
$ clj-kondo --lint repro.clj 
repro.clj:10:1: error: repro/compo is called with 1 arg but expects 2 or 3
linting took 32ms, errors: 1, warnings: 0
```

**config**

```clojure
;; config.clj
{:hooks {:analyze-call {rum.core/defc  hooks.rum/defc
                        rum.core/defcs hooks.rum/defcs}}}
```

`rum.clj` contains the same code as the current master version.

**expected behavior**

No error should be raised.

I tried to change the code by myself but I’m not sure how I should update the `loop`; could you give me any pointer? You may already have some code to parse multi-arity `defn`-like macros. Thanks!

**Comment by borkdude** [comment]
The loop pertains to only one body. A multi-arity function has multiple bodies, so the loop applies to each body. There is code in clj-kondo itself which loops over each element in defn. If a vector is encountered, then we know that we are dealing with a single-arity defn. If a list is encountered, then we know it is a multi-arity function, so we apply the logic to each element in the list and then put it back together again.

**Comment by borkdude** [comment]
@bfontaine Configs from now on are going into this repo: https://github.com/clj-kondo/config

**Comment by bfontaine** [comment]
Thanks! Should we close this issue? I did saw your comment, but haven’t ~~had~~ taken the time to look into the issue for now.

**Comment by borkdude** [comment]
I think we should wait to close this issue when it's fixed over there.

**Comment by borkdude** [comment]
@bfontaine I'm having a stab at this. One question: how can you distinguish a function body from a mixin?

E.g. from the Rum docs:

``` 
(rum/defcs component
  < rum/static
    rum/reactive
    (rum/local 0 ::count)
    (rum/local "" ::text)
  [state label]
  (let [count-atom (::count state)
        text-atom  (::text state)]
    [:div])
``` 

If you write this as a multi-arity:

``` 
(rum/defcs component
  < rum/static
  rum/reactive
  (rum/local 0 ::count)
  (rum/local "" ::text)
  ([state label]
   (component state label nil))
  ([state label _foo]
   (let [count-atom (::count state)
         text-atom  (::text state)]
     [:div])))
``` 

How does the macro know what's a mixin and what is a function body? By looking at the first element of the seq maybe?

**Comment by borkdude** [comment]
I guess so:

https://github.com/tonsky/rum/blob/d60d25a1a3d1c6accfb163e4a8650f7e4ca9255b/src/rum/core.clj#L19-L20

**Comment by borkdude** [comment]
@bfontaine Fixed with https://github.com/clj-kondo/config/commit/60226ba216dabd0399256c771cd771c7160c64a5

**Comment by bfontaine** [comment]
Wooow thank you! :heart_eyes:
